official/keras_application_models/model_callbacks.py

```diff
@@ -0,0 +1,168 @@
+
+import tensorflow as tf  # pylint: disable=g-bad-import-order
+
+from official.utils.logs import logger
+
+# Metrics to log after each batch and epoch
+_PER_BATCH_METRICS = {
+    "loss": "train_loss",
+    "acc": "train_accuracy",
+}
+_PER_EPOCH_METRICS = {
+    "loss": "train_loss",
+    "acc": "train_accuracy",
+    "val_loss": "loss",
+    "val_acc": "accuracy"
+}
+
+
+class ExamplesPerSecondCallback(tf.keras.callbacks.Callback):
+  """ExamplesPerSecond callback.
+
+  This callback records the average_examples_per_sec and
+  current_examples_per_sec during training.
+  """
+
+  def __init__(self, batch_size, every_n_steps=1, metric_logger=None):
+    self._batch_size = batch_size
+    self._every_n_steps = every_n_steps
+    self._logger = metric_logger or logger.BaseBenchmarkLogger()
+    self._global_step = 0  # Initialize it in __init__
+    super(ExamplesPerSecondCallback, self).__init__()
+
+  def on_train_begin(self, logs=None):
+    self._train_start_time = time.time()
+    self._last_recorded_time = time.time()
+
+  def on_batch_begin(self, batch, logs=None):
+    self._time_start = time.time()
+
+  def on_batch_end(self, batch, logs=None):
+    """Log the examples_per_sec metric every_n_steps."""
+    self._global_step += 1
+    current_time = time.time()
+
+    if self._global_step % self._every_n_steps == 0:
+      average_examples_per_sec = self._batch_size * (
+          self._global_step / (current_time - self._train_start_time))
+      self._logger.log_metric(
+          "average_examples_per_sec", average_examples_per_sec,
+          global_step=self._global_step)
+
+      current_examples_per_sec = self._batch_size * (
+          self._every_n_steps / (current_time - self._last_recorded_time))
+      self._logger.log_metric(
+          "current_examples_per_sec", current_examples_per_sec,
+          global_step=self._global_step)
+      self._last_recorded_time = current_time  # Update last_recorded_time
+
+
+class LoggingMetricCallback(tf.keras.callbacks.Callback):
+  """LoggingMetric callback.
+
+  Log the predefined _PER_BATCH_METRICS after each batch, and log the predefined
+  _PER_EPOCH_METRICS after each epoch.
+  """
+
+  def __init__(self, metric_logger=None):
+    self._logger = metric_logger or logger.BaseBenchmarkLogger()
+    self._per_batch_metrics = _PER_BATCH_METRICS
+    self._per_epoch_metrics = _PER_EPOCH_METRICS
+    self._global_step = 0  # Initialize it in __init__
+    super(LoggingMetricCallback, self).__init__()
+
+  def on_batch_end(self, batch, logs=None):
+    """Log metrics after each batch."""
+    self._global_step += 1
+    for metric in _PER_BATCH_METRICS:
+      self._logger.log_metric(
+          _PER_BATCH_METRICS[metric],
+          logs.get(metric),
+          global_step=self._global_step)
+
+  def on_epoch_end(self, epoch, logs=None):
+    """Log metrics after each epoch."""
+    for metric in _PER_EPOCH_METRICS:
+      self._logger.log_metric(
+          _PER_EPOCH_METRICS[metric],
```

Comment:
For the epoch based metric, there are two of them same as the ones in batch based. Should we dedup them since they will be logged by batch based anyway?

Reply:
I think it's okay to keep it. As the number of epochs are not large compared to the number of steps, there won't be too much redundant info.